presto-native-execution/presto_cpp/main/PrestoExchangeSource.cpp

```diff
@@ -195,42 +202,155 @@ void PrestoExchangeSource::doRequest(
-          self->processDataError(
-              path,
-              maxBytes,
-              maxWait,
-              fmt::format(
-                  "Received HTTP {} {} {}",
-                  headers->getStatusCode(),
-                  headers->getStatusMessage(),
-                  bodyAsString(
-                      *response,
-                      self->immediateBufferTransfer_ ? self->pool_.get()
-                                                     : nullptr)));
-        } else if (response->hasError()) {
-          self->processDataError(path, maxBytes, maxWait, response->error());
-        } else {
-          self->processDataResponse(std::move(response));
-        }
-      })
-      .thenError(
-          folly::tag_t<std::exception>{},
-          [path, maxBytes, maxWait, self](const std::exception& e) {
-            self->processDataError(path, maxBytes, maxWait, e.what());
+      .thenTry(
+          [this, path, maxBytes, maxWait, self = getSelfPtr()](
+              folly::Try<std::unique_ptr<http::HttpResponse>> responseTry) {
+            // self needs to be held for keeping 'this' source alive during
+            // processing
+            handleHttpResponse(
+                HttpEndPoint::GET_RESULT,
+                std::move(responseTry),
+                maxWait,
+                maxBytes,
+                path);
           });
 };
 
+void PrestoExchangeSource::acknowledgeResults(int64_t ackSequence) {
+  auto ackPath = fmt::format("{}/{}/acknowledge", basePath_, ackSequence);
+  VLOG(1) << "Sending ack " << ackPath;
+  http::RequestBuilder()
+      .method(proxygen::HTTPMethod::GET)
+      .url(ackPath)
+      .send(httpClient_.get())
+      .via(driverExecutor_)
+      .thenTry(
+          [this, self = getSelfPtr()](
+              folly::Try<std::unique_ptr<http::HttpResponse>> responseTry) {
+            // self needs to be held for keeping 'this' source alive during
+            // processing
+            handleHttpResponse(
+                HttpEndPoint::ACK_RESULT, std::move(responseTry));
+          });
+}
+
+void PrestoExchangeSource::abortResults() {
+  if (abortResultsIssued_.exchange(true)) {
+    return;
+  }
+
+  abortRetryState_ =
+      RetryState(std::chrono::duration_cast<std::chrono::milliseconds>(
+                     SystemConfig::instance()->exchangeMaxErrorDuration())
+                     .count());
+  VLOG(1) << "Sending abort results " << basePath_;
+  doAbortResults(abortRetryState_.nextDelayMs());
+}
+
+void PrestoExchangeSource::doAbortResults(int64_t delayMs) {
+  http::RequestBuilder()
+      .method(proxygen::HTTPMethod::DELETE)
+      .url(basePath_)
+      .send(httpClient_.get(), "", delayMs)
+      .via(driverExecutor_)
+      .thenTry([this, self = getSelfPtr()](
+                   folly::Try<std::unique_ptr<http::HttpResponse>> response) {
+        handleHttpResponse(HttpEndPoint::ABORT_RESULT, std::move(response));
+      });
+}
+
+void PrestoExchangeSource::handleHttpResponse(
```

Comment:
I think different response can have different handler doesn't have to be put together?